In [2]:
import pandas as pd
import numpy as np
import requests

from pandas.io.json import json_normalize

!pip install lxml
import lxml

     |████████████████████████████████| 5.8MB 28.1MB/s eta 0:00:01


#### Reading UN data
cities15000m.csv is a comma separated values file exported from the UN open database [available here](https://unstats.un.org/unsd/demographic-social/products/dyb/dyb_2017/). 

In [3]:
# reading UN data into Pandas dataframe 
c = pd.read_csv('cities15000m.csv',sep=';')
c.head()

geonameid          name  latitude  longitude country code  population  \
0    2650225     Edinburgh  55.95206   -3.19640           GB      464990   
1    2648579       Glasgow  55.86515   -4.25763           GB      591620   
2    1841603      Gyeongju  35.84278  129.21167           KR      155237   
3    4726206   San Antonio  29.42412  -98.49363           US     1469845   
4    3040051  les Escaldes  42.50729    1.53414           AD       15853   

            timezone  
0      Europe/London  
1      Europe/London  
2         Asia/Seoul  
3  \tAmerica/Chicago  
4     Europe/Andorra

In [4]:
# just playing around with data, looking up what's inside
c.loc[c['country code']=='SE'].loc[c.name.str.contains('G')]

geonameid           name  latitude  longitude country code  population  \
18318    2709628    Gustavsberg  59.32684   18.38975           SE       20774   
18319    2711537      Goeteborg  57.70716   11.96679           SE      572799   
18320    2712414         Gaevle  60.67452   17.14174           SE       74884   
18321    2712993  Gamla Uppsala  59.89817   17.63386           SE       16231   

               timezone  
18318  Europe/Stockholm  
18319  Europe/Stockholm  
18320  Europe/Stockholm  
18321  Europe/Stockholm

In [5]:
# just playing around with data, looking up what's inside
c.loc[c['country code']=='CN'].loc[c.name.str.contains('L')].loc[c.name.str.contains('ou')]

geonameid         name  latitude  longitude country code  population  \
3209    1792516  Loushanguan  28.13680  106.82200           CN       80344   
3358    1802238        Loudi  27.73444  111.99444           CN      150684   
3382    1803841     Lianzhou  24.78186  112.37120           CN       92827   
3383    1803842     Lianzhou  21.66621  109.20116           CN       71423   
3393    1804386     Laohekou  32.38583  111.66778           CN      253112   
3394    1804430      Lanzhou  36.05701  103.83987           CN     2628426   
3715    2036066       Linkou  45.28606  130.26151           CN       77754   
3724    2036241      Langtou  40.02611  124.32861           CN       59046   

           timezone  
3209  Asia/Shanghai  
3358  Asia/Shanghai  
3382  Asia/Shanghai  
3383  Asia/Shanghai  
3393  Asia/Shanghai  
3394  Asia/Shanghai  
3715  Asia/Shanghai  
3724  Asia/Shanghai

#### Adjustments
We are going to join this dataframe to our mugs database later. Since datasets come from different sources, some adjustments have to be made. While playing around with data I figured most of the cities that exist in both datasets under different names, and made some adjustment to have them join correctly

In [10]:
# Washington DC adjustment - remove dots
c.loc[c.name.str.contains('D.C.'), 'name'] = 'Washington DC'

# Chiang Mai not joining for some reason
c.loc[c.name.str.contains('Chiang Mai'), 'name'] = 'Chiang Mai' 

In [7]:
# Let's see what are the smallest cities on the set
c.sort_values('population',ascending=False).tail(100)

geonameid                   name  latitude  longitude country code  \
11087    1277202                 Bantva  21.48815   70.07576           IN   
17733     575457               Boguchar  49.93577   40.54500           RU   
22095    5315062               Somerton  32.59644 -114.70968           US   
16244    1180454             Darya Khan  31.79123   71.10394           PK   
20176    4175296            Three Lakes  25.64205  -80.39839           US   
...          ...                    ...       ...        ...          ...   
5374     7302786           Haselbachtal  51.23570   14.02576           DE   
16796    7303944               Melekeok   7.50043  134.62355           PW   
916      2800461          Chasse Royale  50.42842    3.95001           BE   
8254     4043547  Guam Government House  13.47191  144.74978           GU   
23410   11205175               Riverlea -26.21190   27.97322           ZA   

       population             timezone  
11087       15051         Asia/Kolkata  
17733       15048        Europe/Moscow  
22095       15048      America/Phoenix  
16244       15048         Asia/Karachi  
20176       15047     America/New_York  
...           ...                  ...  
5374            0        Europe/Berlin  
16796           0        Pacific/Palau  
916             0      Europe/Brussels  
8254            0         Pacific/Guam  
23410           0  Africa/Johannesburg  

[100 rows x 7 columns]

All the cities with population of 15,000 or less seem to be included in data set. That suits our purpose very well 

In [11]:
# lets create a sorted version of the dataset. cs stands for "cities sorted"
cs = c.sort_values('population', ascending=False)

In [12]:
cs.columns

Index(['geonameid', 'name', 'latitude', 'longitude', 'country code',
       'population', 'timezone'],
      dtype='object')

In [13]:
# rename 'name' column to 'city'
cs.columns = ['geonameid', 'city', 'latitude', 'longitude', 'country code',
       'population', 'timezone']
cs.head()

geonameid          city  latitude  longitude country code  population  \
3266     1796236      Shanghai  31.22222  121.45806           CN    22315474   
19289     745044      Istanbul  41.01384   28.94966           TR    14804116   
194      3435910  Buenos Aires -34.61315  -58.37723           AR    13076300   
10909    1275339        Mumbai  19.07283   72.88261           IN    12691836   
14207    3530597   Mexico City  19.42847  -99.12766           MX    12294193   

                             timezone  
3266                    Asia/Shanghai  
19289                 Europe/Istanbul  
194    America/Argentina/Buenos_Aires  
10909                    Asia/Kolkata  
14207             America/Mexico_City

#### Filtering out duplicate cities
There are some cities with the same name in different countries and even inside some countries, for example:

In [15]:
cs[cs.city=='Albany']

geonameid    city  latitude  longitude country code  population  \
21626    5106834  Albany  42.65258  -73.75623           US       98469   
20201    4179320  Albany  31.57851  -84.15574           US       74843   
22601    5710756  Albany  44.63651 -123.10593           US       52175   
395      2077963  Albany -35.02692  117.88369           AU       33650   
22108    5322850  Albany  37.88687 -122.29775           US       19735   

                  timezone  
21626     America/New_York  
20201     America/New_York  
22601  America/Los_Angeles  
395        Australia/Perth  
22108  America/Los_Angeles

For the purpose of our research we will only take the most populous city among all of the cities with the same name  

In [25]:
# Let's greate a grouped version of the dataset. Since it already sorted by population, we will take the first city from each group 
cg = cs.groupby(cs['city']).first()

In [27]:
cg.loc[cg.index == 'Albany']

geonameid  latitude  longitude country code  population  \
city                                                              
Albany    5106834  42.65258  -73.75623           US       98469   

                timezone  
city                      
Albany  America/New_York

In [31]:
# size before grouping
cs.shape

(23466, 7)

In [32]:
# size after removing duplicates
cg.shape

(22245, 6)

#### Country names
Now we want to decode country codes to actual country names. Otherwise we wouldn't be able to join this table to the mugs dataset  

In [33]:
# read the country codes table from geonames.org 
cn = pd.read_html('https://www.geonames.org/countries/')[1]

In [34]:
cn.head()

ISO-3166alpha2 ISO-3166alpha3  ISO-3166numeric fips               Country  \
0             AD            AND               20   AN               Andorra   
1             AE            ARE              784   AE  United Arab Emirates   
2             AF            AFG                4   AF           Afghanistan   
3             AG            ATG               28   AC   Antigua and Barbuda   
4             AI            AIA              660   AV              Anguilla   

            Capital  Area in km²  Population Continent  
0  Andorra la Vella        468.0       84000        EU  
1         Abu Dhabi      82880.0     4975593        AS  
2             Kabul     647500.0    29121286        AS  
3        St. John's        443.0       86754       NaN  
4        The Valley        102.0       13254       NaN

Let's determine which of the columns we need to use to join dataframes

In [37]:
cn.columns
cn.columns = ['code', 'ISO-3166alpha3', 'ISO-3166numeric', 'fips',
       'Country', 'Capital', 'Area in km²', 'Population', 'Continent']

In [38]:
cn.loc[cn['Country'].str.contains('Mala')] 

code ISO-3166alpha3  ISO-3166numeric fips   Country       Capital  \
157   MW            MWI              454   MI    Malawi      Lilongwe   
159   MY            MYS              458   MY  Malaysia  Kuala Lumpur   

     Area in km²  Population Continent  
157     118480.0    17563749        AF  
159     329750.0    28274729        AS

In [39]:
cg[cg.index=="Kuala Lumpur"]

geonameid  latitude  longitude country code  population  \
city                                                                    
Kuala Lumpur    1735161    3.1412  101.68653           MY     1453975   

                       timezone  
city                             
Kuala Lumpur  Asia/Kuala_Lumpur

'code' column is the one we need, let's set it is dataframe index

In [40]:
cn.index = cn.code

In [20]:
cn.head()

code ISO-3166alpha3  ISO-3166numeric fips               Country  \
code                                                                   
AD     AD            AND               20   AN               Andorra   
AE     AE            ARE              784   AE  United Arab Emirates   
AF     AF            AFG                4   AF           Afghanistan   
AG     AG            ATG               28   AC   Antigua and Barbuda   
AI     AI            AIA              660   AV              Anguilla   

               Capital  Area in km²  Population Continent  
code                                                       
AD    Andorra la Vella        468.0       84000        EU  
AE           Abu Dhabi      82880.0     4975593        AS  
AF               Kabul     647500.0    29121286        AS  
AG          St. John's        443.0       86754       NaN  
AI          The Valley        102.0       13254       NaN

It was found out later, Czech republic is spelled differently in the cities and mugs datasets. We'll adjust it now, before datasets are joined 

In [44]:
cn[cn.index=='CZ']

code ISO-3166alpha3  ISO-3166numeric fips  Country Capital  Area in km²  \
code                                                                           
CZ     CZ            CZE              203   EZ  Czechia  Prague      78866.0   

      Population Continent  
code                        
CZ      10476000        EU

In [47]:
# Czech Republic adjustment
cn.loc[cn['code']=='CZ','Country'] = 'Czech Republic'

In [77]:
# Let's create a joined dataframe
cities = cg.join(cn[['Country','Capital','Population','Continent']], on='country code').reset_index()

In [78]:
cities.head()

city  geonameid  latitude  longitude country code  population  \
0        'Afak      99738  32.06430   45.24743           IQ       21888   
1    'Ain Abid    2508309  36.23194    6.94333           DZ       29486   
2  'Ain Benian    2508275  36.80277    2.92185           DZ       31102   
3   'Ain Deheb    2508225  34.84218    1.54697           DZ       36146   
4  'Ain Merane    2508010  36.16277    0.97037           DZ       34413   

         timezone  Country  Capital  Population Continent  
0    Asia/Baghdad     Iraq  Baghdad    29671605        AS  
1  Africa/Algiers  Algeria  Algiers    34586184        AF  
2  Africa/Algiers  Algeria  Algiers    34586184        AF  
3  Africa/Algiers  Algeria  Algiers    34586184        AF  
4  Africa/Algiers  Algeria  Algiers    34586184        AF

In [79]:
cities.columns

Index(['city', 'geonameid', 'latitude', 'longitude', 'country code',
       'population', 'timezone', 'Country', 'Capital', 'Population',
       'Continent'],
      dtype='object')

In [80]:
# rename Country to country and update index to city and country
cities.columns = ['city','geonameid', 'latitude', 'longitude', 'country code',
       'population', 'timezone', 'country', 'Capital', 'Population', 'Continent']
cities = cities.set_index(['city','country'])

In [27]:
c2.head()

geonameid  latitude  longitude country code  population  \
city        country                                                            
'Afak       Iraq         99738  32.06430   45.24743           IQ       21888   
'Ain Abid   Algeria    2508309  36.23194    6.94333           DZ       29486   
'Ain Benian Algeria    2508275  36.80277    2.92185           DZ       31102   
'Ain Deheb  Algeria    2508225  34.84218    1.54697           DZ       36146   
'Ain Merane Algeria    2508010  36.16277    0.97037           DZ       34413   

                           timezone  Capital  Population Continent  
city        country                                                 
'Afak       Iraq       Asia/Baghdad  Baghdad    29671605        AS  
'Ain Abid   Algeria  Africa/Algiers  Algiers    34586184        AF  
'Ain Benian Algeria  Africa/Algiers  Algiers    34586184        AF  
'Ain Deheb  Algeria  Africa/Algiers  Algiers    34586184        AF  
'Ain Merane Algeria  Africa/Algiers  Algiers    34586184        AF

#### Mugs dataset

In [84]:
# Read mugs data
m = pd.read_json('mugs.json', orient='index')
m = m.reset_index().drop(columns=['badge','icon','image','imageBy'])
m.columns = ['id', 'city', 'country', 'month', 'name', 'series', 'state', 'year'] 
m.head()

id       city               country    month       name series  \
0     aachen_icon     Aachen               Germany  January     Aachen   icon   
1     aarhus_icon     Aarhus               Denmark  January     Aarhus   icon   
2    aberdeen_yah   Aberdeen        United Kingdom   August   Aberdeen    yah   
3  abu_dhabi_icon  Abu Dhabi  United Arab Emirates  January  Abu Dhabi   icon   
4    acapulco_bts   Acapulco                Mexico   August   Acapulco    bts   

      state  year  
0       NaN  2009  
1       NaN  2014  
2  Scotland  2017  
3       NaN  2011  
4  Guerrero  2018

Add a column with release date - month and year combined and converted to date, so that we could sort the dataframe by release date

In [85]:
from datetime import datetime

#strptime sample
dstr = '{} {}'.format('January', 2018)
dt = datetime.strptime(dstr, '%B %Y')
print(dt)

2018-01-01 00:00:00


In [86]:
# Add Date column
m['date'] = (m['month'] + ' ' + m['year'].astype(str)).apply(lambda x: datetime.strptime(x, '%B %Y'))

In [87]:
m.head()

id       city               country    month       name series  \
0     aachen_icon     Aachen               Germany  January     Aachen   icon   
1     aarhus_icon     Aarhus               Denmark  January     Aarhus   icon   
2    aberdeen_yah   Aberdeen        United Kingdom   August   Aberdeen    yah   
3  abu_dhabi_icon  Abu Dhabi  United Arab Emirates  January  Abu Dhabi   icon   
4    acapulco_bts   Acapulco                Mexico   August   Acapulco    bts   

      state  year       date  
0       NaN  2009 2009-01-01  
1       NaN  2014 2014-01-01  
2  Scotland  2017 2017-08-01  
3       NaN  2011 2011-01-01  
4  Guerrero  2018 2018-08-01

In [89]:
# check if sorting by date works
m.sort_values('date', ascending=False).head()

id      city            country    month  \
708              luzhou_yah    Luzhou              China  October   
785            mianyang_yah  Mianyang              China  October   
1426            zhuzhou_yah   Zhuzhou              China  October   
1294  trinidad_&_tobago_bts       NaN  Trinidad & Tobago  October   
660              leshan_yah    Leshan              China  October   

                   name series state  year       date  
708              Luzhou    yah   NaN  2019 2019-10-01  
785            Mianyang    yah   NaN  2019 2019-10-01  
1426            Zhuzhou    yah   NaN  2019 2019-10-01  
1294  Trinidad & Tobago    bts   NaN  2019 2019-10-01  
660              Leshan    yah   NaN  2019 2019-10-01

#### Adjustments
Again, since we need to join two different datasources, some adjustments need to be made to city names spellings. I will add a new column, cun that stands for City-United-Nations spelling. All these discrepancies were figured out by playing with data. Specifically, after joining datasets I went through .isnull() values to figure out which cities from mugs dataset were not found in the UN dataset.

In [90]:
# City names adjustments
m['cun'] = m['city']
m.loc[m.cun == 'Washington', 'cun'] = 'Washington DC'
m.loc[m.cun == 'Munster', 'cun'] = 'Muenster'
m.loc[m.cun == 'Aarhus', 'cun'] = 'Arhus'
m.loc[m.cun == 'Acapulco', 'cun'] = 'Acapulco de Juarez'
m.loc[m.cun == 'Antigua', 'cun'] = 'Antigua Guatemala'
m.loc[m.cun == 'Antwerp', 'cun'] = 'Antwerpen'
m.loc[m.cun == 'Bacolod', 'cun'] = 'Bacolod City'
m.loc[m.cun == 'Berne', 'cun'] = 'Bern'
m.loc[m.cun == 'Brunswick', 'cun'] = 'Braunschweig'
m.loc[m.cun == 'Cebu', 'cun'] = 'Cebu City'
m.loc[m.cun == 'Chiangmai', 'cun'] = 'Сhiang Mai'
m.loc[m.cun == 'Cologne', 'cun'] = 'Koeln'
m.loc[m.cun == 'Dusseldorf', 'cun'] = 'Duesseldorf'
m.loc[m.cun == 'Geneva', 'cun'] = 'Geneve'
m.loc[m.cun == 'Gothenburg', 'cun'] = 'Goeteborg'
m.loc[m.cun == 'Hai Phong', 'cun'] = 'Haiphong'
m.loc[m.cun == 'Hong Kong', 'country'] = 'Hong Kong'
m.loc[m.cun == 'Jeju', 'cun'] = 'Jeju City'
m.loc[m.cun == 'Leon', 'cun'] = 'Leon de los Aldama'
m.loc[m.cun == 'Lucerne', 'cun'] = 'Luzern'
m.loc[m.cun == 'Macau', 'country'] = 'Macao'
m.loc[m.cun == 'Makati', 'cun'] = 'Makati City'
m.loc[m.cun == 'Malmo', 'cun'] = 'Malmoe'
m.loc[m.cun == 'Marrakech', 'cun'] = 'Marrakesh'
m.loc[m.cun == 'New York', 'cun'] = 'New York City'
m.loc[m.cun == 'Newcastle Upon Tyne', 'cun'] = 'Newcastle upon Tyne'
m.loc[m.cun == 'Nuremberg', 'cun'] = 'Nuernberg'
m.loc[m.cun == 'Pattaya', 'cun'] = 'Phatthaya'
m.loc[m.cun == 'Playa Del Carmen', 'cun'] = 'Playa del Carmen'
m.loc[m.cun == 'Ras Al Khaimah', 'cun'] = 'Ras al-Khaimah'
m.loc[m.cun == 'Rio De Janeiro', 'cun'] = 'Rio de Janeiro'
m.loc[m.cun == 'Rostov-on-Don', 'cun'] = 'Rostov-na-Donu'
m.loc[m.cun == 'Seville', 'cun'] = 'Sevilla'
m.loc[m.cun == 'St. Gallen', 'cun'] = 'Sankt Gallen'
m.loc[m.cun == 'St. Petersburg', 'cun'] = 'Saint Petersburg'
m.loc[m.cun == 'Taoyuan', 'cun'] = 'Taoyuan City'
m.loc[m.cun == 'Tsingtao', 'cun'] = 'Qingdao'
m.loc[m.cun == 'Xian', 'cun'] = 'Xi\'an'

In [92]:
# Let's join the datasets. mc statnds for mugs and cities
mc = m.join(cities, on=['cun','country'])

In [93]:
mc.shape

(1431, 19)

#### Joined dataset
For the first time in history (lol) we have created a dataset like this. All of existing Strabucks mugs are now listed next to their cities population values. Let's quickly get some insights 

In [95]:
# Now we can easily show the least populous cities that have Starbucks mugs 
mc[['city','country','population']].groupby('city').first().sort_values('population').head()

country  population
city                             
Corfu          Greece     27003.0
Bodrum         Turkey     39317.0
Antigua     Guatemala     39368.0
Roermond  Netherlands     44975.0
Lucerne   Switzerland     57066.0

In [98]:
# Let's create a grouped data set of the cities that Starbucks mugs. (Before grouping there might be several mugs per city in the mc dataset)
mg = mc.groupby(['city','country']).first()

In [102]:
# Let's show the cities that was now found in UN dataset
print(mg.loc[mg['population'].isnull()].shape)
mg.loc[mg['population'].isnull()].head(20).tail(10)

(37, 17)


id     month  \
city             country                                                  
Lake Buena Vista United States  disneys_hollywood studios_yah   January   
Langkawi         Malaysia                        langkawi_yah  November   
Liuzhou          China                           liuzhou_icon   January   
Los Cabos        Mexico                         los_cabos_bts    August   
Luzhou           China                             luzhou_yah   October   
Ma'anshan        China                           maanshan_yah  November   
Marne la Vallee  France                  disneyland_paris_yah      June   
Mykonos          Greece                          mykonos_icon   January   
Orange County    United States              orange_county_yah     March   
Pampanga         Philippines                    pampanga_icon   January   

                                                      name series  \
city             country                                            
Lake Buena Vista United States  Disney's Hollywood Studios    yah   
Langkawi         Malaysia                         Langkawi    yah   
Liuzhou          China                             Liuzhou   icon   
Los Cabos        Mexico                          Los Cabos    bts   
Luzhou           China                              Luzhou    yah   
Ma'anshan        China                           Ma'anshan    yah   
Marne la Vallee  France                   Disneyland Paris    yah   
Mykonos          Greece                            Mykonos   icon   
Orange County    United States               Orange County    yah   
Pampanga         Philippines                      Pampanga   icon   

                                          state  year       date  \
city             country                                           
Lake Buena Vista United States          Florida  2015 2015-01-01   
Langkawi         Malaysia                   NaN  2018 2018-11-01   
Liuzhou          China                      NaN  2013 2013-01-01   
Los Cabos        Mexico         Baja California  2018 2018-08-01   
Luzhou           China                      NaN  2019 2019-10-01   
Ma'anshan        China                      NaN  2018 2018-11-01   
Marne la Vallee  France                     NaN  2018 2018-06-01   
Mykonos          Greece                     NaN  2011 2011-01-01   
Orange County    United States       California  2013 2013-03-01   
Pampanga         Philippines                NaN  2012 2012-01-01   

                                             cun  geonameid  latitude  \
city             country                                                
Lake Buena Vista United States  Lake Buena Vista        NaN       NaN   
Langkawi         Malaysia               Langkawi        NaN       NaN   
Liuzhou          China                   Liuzhou        NaN       NaN   
Los Cabos        Mexico                Los Cabos        NaN       NaN   
Luzhou           China                    Luzhou        NaN       NaN   
Ma'anshan        China                 Ma'anshan        NaN       NaN   
Marne la Vallee  France          Marne la Vallee        NaN       NaN   
Mykonos          Greece                  Mykonos        NaN       NaN   
Orange County    United States     Orange County        NaN       NaN   
Pampanga         Philippines            Pampanga        NaN       NaN   

                                longitude country code  population timezone  \
city             country                                                      
Lake Buena Vista United States        NaN          NaN         NaN      NaN   
Langkawi         Malaysia             NaN          NaN         NaN      NaN   
Liuzhou          China                NaN          NaN         NaN      NaN   
Los Cabos        Mexico               NaN          NaN         NaN      NaN   
Luzhou           China                NaN          NaN         NaN      NaN   
Ma'anshan        China                NaN          NaN         NaN      NaN   
Mar

The table above was used to adjust existing city names.
All the Null-values left in the table are either special locations such as Islands, States, Countries, or Disneylands, or they indeed was not found in the UN dataset

In [105]:
cities.shape

(22245, 9)

We will need to query Foursquare for each of our 22,245 records, if that city has a Starbucks shop. Since there is a limit to Foursqare requests per hour, lets shrink this dataset and keep only the cities larger than 30,000 of population 

In [106]:
cities.loc[cities['population']>30000].shape

(13110, 9)

In [107]:
c3 = cities.loc[cities['population']>30000].drop(columns=['timezone', 'Capital', 'Population', 'Continent']).reset_index()

In [116]:
print(c3.shape)
c3.head()

(13110, 8)


city  country  geonameid  latitude  longitude country code  \
0   'Ain Benian  Algeria    2508275  36.80277    2.92185           DZ   
1    'Ain Deheb  Algeria    2508225  34.84218    1.54697           DZ   
2   'Ain Merane  Algeria    2508010  36.16277    0.97037           DZ   
3  'Ain el Bell  Algeria    2508184  34.34381    3.22475           DZ   
4  'Ain el Melh  Algeria    2508130  34.84146    4.16383           DZ   

   population  hasMug  
0       31102   False  
1       36146   False  
2       34413   False  
3       31916   False  
4       39798   False

#### hasMug field
Let's adda column to the UN cities dataset, that would tell us if that city has a Starbucks mug

In [113]:
# to do so, we'll define a simple function
m3 = mg.reset_index()

def hasMug(city: str, country: str):
    return not m3.loc[m3.country==country].loc[m3.city==city].empty

In [114]:
# testing hasMug function
hasMug('New York City','United States')

True

In [115]:
# applying the function to the dataset
c3['hasMug'] = c3[['city','country']].apply(lambda x: hasMug(x['city'],x['country']), axis=1 )

In [117]:
c3.head()

city  country  geonameid  latitude  longitude country code  \
0   'Ain Benian  Algeria    2508275  36.80277    2.92185           DZ   
1    'Ain Deheb  Algeria    2508225  34.84218    1.54697           DZ   
2   'Ain Merane  Algeria    2508010  36.16277    0.97037           DZ   
3  'Ain el Bell  Algeria    2508184  34.34381    3.22475           DZ   
4  'Ain el Melh  Algeria    2508130  34.84146    4.16383           DZ   

   population  hasMug  
0       31102   False  
1       36146   False  
2       34413   False  
3       31916   False  
4       39798   False

In [118]:
# we can now filter out the cities that have or don't have Starbucks mug
c3.loc[c3.hasMug]

city               country  geonameid  latitude  longitude  \
12        Aachen               Germany    3247449  50.77664    6.08342   
33      Aberdeen        United Kingdom    2657832  57.14369   -2.09814   
49     Abu Dhabi  United Arab Emirates     292968  24.46667   54.36667   
85         Adana                Turkey     325363  37.00167   35.32889   
346     Alicante                 Spain    2521978  38.34517   -0.48149   
...          ...                   ...        ...       ...        ...   
13018  Zhongshan                 China    1915223  21.31992  110.57230   
13030     Zhuhai                 China    1790437  22.27694  113.56778   
13036    Zhuzhou                 China    1783763  27.83333  113.15000   
13075      Zunyi                 China    1783621  27.68667  106.90722   
13076     Zurich           Switzerland    2657896  47.36667    8.55000   

      country code  population  hasMug  
12              DE      265208    True  
33              GB      196670    True  
49              AE      603492    True  
85              TR     1248988    True  
346             ES      334757    True  
...            ...         ...     ...  
13018           CN     3121275    True  
13030           CN      501199    True  
13036           CN      709358    True  
13075           CN      466292    True  
13076           CH      341730    True  

[329 rows x 8 columns]

For some reason UN dataset contains Delhi and New Delhi as separate records. We will simply remove Delhi at a later stage, to have our results cleaner

In [122]:
# This a check for Delhi. There is a New Delhi Starbucks mug, but no Delhi mug. 
c3[c3['city'].str.contains('Delhi')]

city country  geonameid  latitude  longitude country code  \
3002      Delhi   India    1273294  28.65195   77.23149           IN   
8022  New Delhi   India    1261481  28.63576   77.22445           IN   

      population  hasMug  
3002    10927986   False  
8022      317797    True

In [123]:
m.head()

id       city               country    month       name series  \
0     aachen_icon     Aachen               Germany  January     Aachen   icon   
1     aarhus_icon     Aarhus               Denmark  January     Aarhus   icon   
2    aberdeen_yah   Aberdeen        United Kingdom   August   Aberdeen    yah   
3  abu_dhabi_icon  Abu Dhabi  United Arab Emirates  January  Abu Dhabi   icon   
4    acapulco_bts   Acapulco                Mexico   August   Acapulco    bts   

      state  year       date                 cun  
0       NaN  2009 2009-01-01              Aachen  
1       NaN  2014 2014-01-01               Arhus  
2  Scotland  2017 2017-08-01            Aberdeen  
3       NaN  2011 2011-01-01           Abu Dhabi  
4  Guerrero  2018 2018-08-01  Acapulco de Juarez

In [120]:
m[['city','state','country']].head()

city     state               country
0     Aachen       NaN               Germany
1     Aarhus       NaN               Denmark
2   Aberdeen  Scotland        United Kingdom
3  Abu Dhabi       NaN  United Arab Emirates
4   Acapulco  Guerrero                Mexico

Now we will define functions to tell location_type (city/state/country/special) and location_path (e.g. New York, United States) for each mug so that we could query geocoder for cities coordinates.

In [125]:
def location_type(city_state_country_name):
    city = city_state_country_name[0]
    state = city_state_country_name[1]
    country = city_state_country_name[2]
    name = city_state_country_name[3]
    
    if 'Disney' in name or 'Universal' in name or 'Magic' in name or 'Epcot' in name or 'Pike' in name:
        return 'special'
    
    location_type = 'country'
    
    if not state != state:
        location_type = 'state'
    
    if not city != city:
        location_type = 'city'
    
    return location_type

def location_path(city_state_country):
    city = city_state_country[0]
    state = city_state_country[1]
    country = city_state_country[2]
    
    path = country
    
    if not state != state:
        path = state + ', ' + path
    
    if not city != city:
        path = city +  ', ' + path
    
    return path

In [126]:
# add location_type column to dataset
m['location_type'] = m[['city','state','country','name']].apply(lambda x: location_type(x), axis = 1)

In [127]:
# add location_path to dataset
m['location_path'] = m[['city','state','country']].apply(lambda x: location_path(x), axis = 1)

Double checking the results

In [128]:
m.head()

id       city               country    month       name series  \
0     aachen_icon     Aachen               Germany  January     Aachen   icon   
1     aarhus_icon     Aarhus               Denmark  January     Aarhus   icon   
2    aberdeen_yah   Aberdeen        United Kingdom   August   Aberdeen    yah   
3  abu_dhabi_icon  Abu Dhabi  United Arab Emirates  January  Abu Dhabi   icon   
4    acapulco_bts   Acapulco                Mexico   August   Acapulco    bts   

      state  year       date                 cun location_type  \
0       NaN  2009 2009-01-01              Aachen          city   
1       NaN  2014 2014-01-01               Arhus          city   
2  Scotland  2017 2017-08-01            Aberdeen          city   
3       NaN  2011 2011-01-01           Abu Dhabi          city   
4  Guerrero  2018 2018-08-01  Acapulco de Juarez          city   

                        location_path  
0                     Aachen, Germany  
1                     Aarhus, Denmark  
2  Aberdeen, Scotland, United Kingdom  
3     Abu Dhabi, United Arab Emirates  
4          Acapulco, Guerrero, Mexico

In [129]:
m[m.name.str.contains('Magic')]

id              city        country      month  \
721     magic_kingdom_bto  Lake Buena Vista  United States       July   
722     magic_kingdom_bts  Lake Buena Vista  United States      April   
723     magic_kingdom_yah  Lake Buena Vista  United States    January   
724   magic_kingdom_yah_2  Lake Buena Vista  United States  September   
725   magic_kingdom_yah_3  Lake Buena Vista  United States        May   
726    magic_kingdom_yaho  Lake Buena Vista  United States     August   
727  magic_kingdom_yaho_2  Lake Buena Vista  United States   November   

                  name series    state  year       date               cun  \
721      Magic Kingdom    bts  Florida  2019 2019-07-01  Lake Buena Vista   
722      Magic Kingdom    bts  Florida  2019 2019-04-01  Lake Buena Vista   
723      Magic Kingdom    yah  Florida  2015 2015-01-01  Lake Buena Vista   
724  Magic Kingdom v.2    yah  Florida  2017 2017-09-01  Lake Buena Vista   
725  Magic Kingdom v.3    yah  Florida  2018 2018-05-01  Lake Buena Vista   
726      Magic Kingdom   yaho  Florida  2016 2016-08-01  Lake Buena Vista   
727  Magic Kingdom v.2   yaho  Florida  2017 2017-11-01  Lake Buena Vista   

    location_type                             location_path  
721       special  Lake Buena Vista, Florida, United States  
722       special  Lake Buena Vista, Florida, United States  
723       special  Lake Buena Vista, Florida, United States  
724       special  Lake Buena Vista, Florida, United States  
725       special  Lake Buena Vista, Florida, United States  
726       special  Lake Buena Vista, Florida, United States  
727       special  Lake Buena Vista, Florida, United States

In [130]:
m['location_type'].value_counts()

city       914
country    232
state      230
special     55
Name: location_type, dtype: int64

#### Locations for cities that do have mugs
Now it's the time to query 

In [303]:
!pip install geopy
from geopy.geocoders import Nominatim

In [327]:
geolocator = Nominatim(user_agent="city_explorer_2")

for index, row in m.iterrows():
    if index > 1215:
        location = geolocator.geocode(row['location_path'])

        if not location is None:
            m.at[index, 'latitude'] = location.latitude
            m.at[index, 'longitude'] = location.longitude


GeocoderQuotaExceeded: HTTP Error 429: Too Many Requests

In [325]:
m.head(1220).tail()

id    city country      month         name series state  year  \
1215  suzhou_icon  Suzhou   China    January       Suzhou   icon   NaN  2011   
1216   suzhou_rlf  Suzhou   China    January  Suzhou Blue    rlf   NaN  2013   
1217   suzhou_yah  Suzhou   China  September       Suzhou    yah   NaN  2017   
1218  sweden_icon     NaN  Sweden    January       Sweden   icon   NaN  2012   
1219   sweden_yah     NaN  Sweden  September       Sweden    yah   NaN  2018   

           date     cun location_type  location_path   latitude   longitude  
1215 2011-01-01  Suzhou          city  Suzhou, China  31.402994  120.733535  
1216 2013-01-01  Suzhou          city  Suzhou, China        NaN         NaN  
1217 2017-09-01  Suzhou          city  Suzhou, China        NaN         NaN  
1218 2012-01-01     NaN       country         Sweden        NaN         NaN  
1219 2018-09-01     NaN       country         Sweden        NaN         NaN

In [326]:
m.to_csv('m_1215.csv', index = False)